File: Protocol/_theoreme_sim/src/core/graphUtils.js
/* Purpose: Generic graph utilities for adjacency, BFS distances, edge lookup, and geometry. */

/**
 * @param {number} a
 * @param {number} b
 * @returns {string}
 */
export function edgeKey(a, b) {
  return a < b ? `${a}:${b}` : `${b}:${a}`;
}

/**
 * @param {{x:number,y:number}} u
 * @param {{x:number,y:number}} v
 * @returns {number}
 */
export function distance(u, v) {
  const dx = u.x - v.x;
  const dy = u.y - v.y;
  return Math.sqrt(dx * dx + dy * dy);
}

/**
 * @param {number} nodeCount
 * @param {{a:number,b:number}[]} edges
 * @returns {Map<number, Set<number>>}
 */
export function buildAdjacency(nodeCount, edges) {
  const adjacency = new Map();
  for (let i = 0; i < nodeCount; i += 1) {
    adjacency.set(i, new Set());
  }

  for (const edge of edges) {
    adjacency.get(edge.a).add(edge.b);
    adjacency.get(edge.b).add(edge.a);
  }

  return adjacency;
}

/**
 * @param {{a:number,b:number}[]} edges
 * @returns {Map<string, {a:number,b:number,distance:number,quality:number}>}
 */
export function buildEdgeLookup(edges) {
  const lookup = new Map();
  for (const edge of edges) {
    lookup.set(edgeKey(edge.a, edge.b), edge);
  }
  return lookup;
}

/**
 * @param {Map<number, Set<number>>} adjacency
 * @returns {number[][]}
 */
export function connectedComponents(adjacency) {
  const seen = new Set();
  const components = [];

  for (const id of adjacency.keys()) {
    if (seen.has(id)) {
      continue;
    }

    const queue = [id];
    const component = [];
    seen.add(id);

    while (queue.length > 0) {
      const current = queue.shift();
      component.push(current);

      for (const next of adjacency.get(current)) {
        if (!seen.has(next)) {
          seen.add(next);
          queue.push(next);
        }
      }
    }

    components.push(component);
  }

  return components;
}

/**
 * @param {Map<number, Set<number>>} adjacency
 * @param {number} rootId
 * @returns {Map<number, number>}
 */
export function bfsDistances(adjacency, rootId) {
  const dist = new Map();
  for (const id of adjacency.keys()) {
    dist.set(id, Number.POSITIVE_INFINITY);
  }

  dist.set(rootId, 0);
  const queue = [rootId];

  while (queue.length > 0) {
    const current = queue.shift();
    const currentDist = dist.get(current);

    for (const next of adjacency.get(current)) {
      if (dist.get(next) === Number.POSITIVE_INFINITY) {
        dist.set(next, currentDist + 1);
        queue.push(next);
      }
    }
  }

  return dist;
}

File: Protocol/_theoreme_sim_py/src/core/graph_utils.py
"""Purpose: Graph helpers for adjacency, BFS distances, and geometry operations."""

from __future__ import annotations

from collections import deque


def edge_key(a: int, b: int) -> str:
    return f"{a}:{b}" if a < b else f"{b}:{a}"


def distance(u: dict, v: dict) -> float:
    dx = float(u["x"]) - float(v["x"])
    dy = float(u["y"]) - float(v["y"])
    return (dx * dx + dy * dy) ** 0.5


def build_adjacency(node_count: int, edges: list[dict]) -> dict[int, set[int]]:
    adjacency: dict[int, set[int]] = {index: set() for index in range(node_count)}
    for edge in edges:
        adjacency[edge["a"]].add(edge["b"])
        adjacency[edge["b"]].add(edge["a"])
    return adjacency


def build_edge_lookup(edges: list[dict]) -> dict[str, dict]:
    lookup: dict[str, dict] = {}
    for edge in edges:
        lookup[edge_key(edge["a"], edge["b"])] = edge
    return lookup


def connected_components(adjacency: dict[int, set[int]]) -> list[list[int]]:
    seen: set[int] = set()
    components: list[list[int]] = []

    for node_id in adjacency:
        if node_id in seen:
            continue
        queue = deque([node_id])
        seen.add(node_id)
        component: list[int] = []

        while queue:
            current = queue.popleft()
            component.append(current)
            for nxt in adjacency[current]:
                if nxt not in seen:
                    seen.add(nxt)
                    queue.append(nxt)

        components.append(component)

    return components


def bfs_distances(adjacency: dict[int, set[int]], root_id: int) -> dict[int, float]:
    dist: dict[int, float] = {
        node_id: float("inf") for node_id in adjacency
    }
    dist[root_id] = 0.0
    queue = deque([root_id])

    while queue:
        current = queue.popleft()
        current_dist = dist[current]
        for nxt in adjacency[current]:
            if dist[nxt] == float("inf"):
                dist[nxt] = current_dist + 1.0
                queue.append(nxt)

    return dist
